### ISSUE TITLE
Serialize and deserialize a projection

### ISSUE THREAD

**Issue by kitsonk** [dojo/widget-core#485]
**Feature**

In support of server side rendering, we need a mechanism that can take a `Projector` and serialize the virtual DOM in a way that it can be restored in another projector.  This allows context switching of the virtual DOM.

Things this feature needs to do technically:

- Provide a plain object structure which does not contain references to functions, DOM node instances, or other non-plain objects.
- Be able to hydrate the maquette projection where upon the first "proper" render, maquette will not remove and re-add the whole DOM tree.
- Allow a subsequent render of the projector add listeners and other references that are generated by the "stood up" application.

**Comment by kitsonk** [comment]
I discovered a few things with the WIP at #484:

- I assumed I would have to populate the `VNode.domNode` for maquette in order to not have it totally discard the DOM tree and properly stitch together the DOM with the virtual DOM.  In reality, I ended up replicating the `createDom` functionality in maquette only to realise that if I provided a `VNode` without a `domNode` that it would go ahead and create it.
- As @matt-gadd and I discussed, the serialisation strips out all functions, including `afterCreate`.  If we have abstractions that need to know about the actual DOM nodes, we are going to have to figure out a mechanism to inform these of the create DOM nodes.
- I _think_ I am stripping out all the dangerous stuff from the serialization, but it will need a fair amount of testing, especially when something is stripped out but is later added in when there is a proper render of the `Projector`.
